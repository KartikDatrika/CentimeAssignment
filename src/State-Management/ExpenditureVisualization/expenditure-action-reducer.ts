import { ADD_FLOW, DELETE_FLOW, EDIT_FLOW, GET_EXPENDITURE_DATA, IExpenditureAction } from "./expenditure-action-types";

const initialData = {
    data: [] as any
}

export const EG_INCOME = "eg: Income";
export const EG_EXPENDITURE = "eg: Expenditure";

export default function expenditureReducer(state = initialData, action: IExpenditureAction) {
    
    let newData = action.payload && [...action.payload.data];
    
    switch (action.type) {
        case GET_EXPENDITURE_DATA:
            break;
        case ADD_FLOW:
            newData.splice(1, 0, [EG_INCOME, EG_EXPENDITURE, 1]);
            break;
        case EDIT_FLOW:
            newData[action.payload.x][action.payload.y] = action.payload.value;
            break;
        case DELETE_FLOW:
            newData.splice(action.payload.x, 1);
            break;
        default:
            return state;
    }
    
    return {
        data: newData
    }
}