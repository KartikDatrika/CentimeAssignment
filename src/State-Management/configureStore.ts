import { applyMiddleware, compose, createStore, combineReducers } from 'redux'
import thunkMiddleware from 'redux-thunk'
import expenditureReducer from './ExpenditureVisualization/expenditure-action-reducer'

export interface IAppState{
    expenditureData: {data:any[]}
}

const rootReducer = combineReducers({
    expenditureData: expenditureReducer
})

export default function configureStore() {
    const middlewares = [thunkMiddleware]
    const middlewareEnhancer = applyMiddleware(...middlewares)
    const composedEnhancers = compose(middlewareEnhancer)
    const store = createStore(rootReducer, composedEnhancers);
    return store
}
