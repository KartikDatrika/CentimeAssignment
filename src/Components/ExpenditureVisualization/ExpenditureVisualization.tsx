import { useEffect } from 'react'
import { ReactGoogleChartProps } from 'react-google-charts/dist/types';
import { useDispatch, useSelector } from 'react-redux';
import { IAppState } from '../../State-Management/configureStore';
import { fetchDataFromServer } from '../../State-Management/ExpenditureVisualization/expenditure-action-creators';
import GoogleChart from '../Charts/GoogleChart';
import LanguageSwitcher from '../Utility/LanguageSwitcher/LanguageSwitcher';
import EditableTable from '../Utility/EditableTable/EditableTable';

export default function ExpenditureVisualization() {

    const data = useSelector((state: IAppState) => state.expenditureData.data);
    const dispatch = useDispatch();

    useEffect(() => {
        fetchDataFromServer(dispatch, 'http://localhost:3001/flows');
    }, [])

    let chartProps: ReactGoogleChartProps = {
        chartType: "Sankey",
        height: 500,
        data
    }

    const chart = data.length > 0 ? <GoogleChart {...chartProps} /> : null;

    return (
        <div className="container">
            <LanguageSwitcher/>
            <EditableTable translateColumnNames={true}/>
            {chart}
        </div>
    )
}
