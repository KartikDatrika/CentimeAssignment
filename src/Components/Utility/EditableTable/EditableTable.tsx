import { useTranslation } from 'react-i18next';
import { useSelector, useDispatch } from 'react-redux';
import { IAppState } from '../../../State-Management/configureStore';
import { addRow, deleteRow, editRow } from '../../../State-Management/ExpenditureVisualization/expenditure-action-creators';
import { EG_EXPENDITURE, EG_INCOME } from '../../../State-Management/ExpenditureVisualization/expenditure-action-reducer';

export default function EditableTable(props: {translateColumnNames: boolean}) {

    const data = useSelector((state: IAppState) => state.expenditureData.data);
    const dispatch = useDispatch();

    const { t } = useTranslation();
    
    const onAddClick = () => {
        addRow(dispatch, data);
    }
    const onEditClick = (rowId: number, columnId: number, value: any) => {
        value = isNaN(Number(value)) ? value : Number(value);
        editRow(dispatch, data, rowId, columnId, value as Number);
    }
    const onDeleteClick = (rowId: number) => {
        deleteRow(dispatch, data, rowId);
    }

    return (

        data && data.length > 0 ?
            <div className={"table-container"}>
                <table className="table table-borderless"> 
                    <thead>
                        <tr>
                            {data[0].map((header: any, index: number) =>
                                <td className="columnName" key={index + "td"}> {props.translateColumnNames ? t(header) : header} </td>
                            )}
                            <td><button className="btn btn-success" key={"add"} type="button" onClick={onAddClick}>ADD</button></td>
                        </tr>
                    </thead>
                    <tbody>
                        {data.slice(1,).map((flow: any[], rowId) => <tr>
                            {flow.map((cell: any, columnId) => 
                                <td>
                                    <input className="form-control" 
                                        value={(cell === EG_INCOME ||  cell === EG_EXPENDITURE) ? "" : cell} 
                                        placeholder={(cell === EG_INCOME || cell === EG_EXPENDITURE) ? cell : ""} 
                                        onChange={(e) => onEditClick(rowId + 1, columnId, e.target.value)}>
                                    </input></td>)}
                                <td>
                                <button className="btn btn-danger" type="button" onClick={() => onDeleteClick(rowId + 1)}>DEL</button>
                            </td>
                        </tr>)}
                    </tbody>
                </table>
            </div> : null
    )
}
