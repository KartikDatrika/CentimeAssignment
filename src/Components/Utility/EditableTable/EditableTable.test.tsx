import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import EditableTable from './EditableTable';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { addRow, deleteRow, editRow } from '../../../State-Management/ExpenditureVisualization/expenditure-action-creators';
jest.mock('../../../State-Management/ExpenditureVisualization/expenditure-action-creators', () => ({
    addRow: jest.fn(),
    deleteRow: jest.fn(),
    editRow: jest.fn()
}));

describe('EditableTable Test Suite', () => {
    let mockStore = {} as any;
    let store: any;
    let response: Response = {
        json: [
            [
                "InFlow",
                "OutFlow",
                "Amount"
            ],
            [
                "Stocks",
                "Savings",
                25
            ],
            [
                "Gold",
                "Savings",
                10
            ]]
    } as any;
    beforeAll(() => {
        mockStore = configureStore();
        store = mockStore({ expenditureData: { data: response.json } });
    })

    it('Editable Table is rendering without error', () => {
        const { container } = render(<Provider store={store}><EditableTable translateColumnNames={false} /></Provider>);
        expect(container).toMatchSnapshot();
    });


    it('Editable Table is rendering with one add and 2 delete buttons.', () => {
        const {getAllByRole } = render(<Provider store={store}><EditableTable translateColumnNames={false} /></Provider>);
        const buttons = getAllByRole("button");
        const addButton = buttons[0];
        const delButton1 = buttons[1];
        const delButton2 = buttons[2];
        expect(addButton.className).toEqual("btn btn-success");
        expect(delButton1.className).toEqual("btn btn-danger");
        expect(delButton2.className).toEqual("btn btn-danger");
    });

    it('on add/delete buttons click expected action creators are triggered.', () => {
        const { getAllByRole } = render(<Provider store={store}><EditableTable translateColumnNames={false} /></Provider>);
        const buttons = getAllByRole("button");
        const addButton = buttons[0];
        addButton.click();
        expect(addRow).toBeCalled();
        const delButton = buttons[1];
        delButton.click();
        expect(deleteRow).toBeCalled();
    });

    it('on changing table cell and edit action creator is triggered.', () => {
        const { getByDisplayValue } = render(<Provider store={store}><EditableTable translateColumnNames={false} /></Provider>);
        const inputCell = getByDisplayValue("Stocks");
        fireEvent.change(inputCell, {target: {value: "1"}});
        expect(editRow).toBeCalled();
    });
})

