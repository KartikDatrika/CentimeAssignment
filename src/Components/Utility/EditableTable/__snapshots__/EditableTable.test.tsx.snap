// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTable Test Suite Editable Table is rendering without error 1`] = `
<div>
  <div
    class="table-container"
  >
    <table
      class="table table-borderless"
    >
      <thead>
        <tr>
          <td
            class="columnName"
          >
             
            InFlow
             
          </td>
          <td
            class="columnName"
          >
             
            OutFlow
             
          </td>
          <td
            class="columnName"
          >
             
            Amount
             
          </td>
          <td>
            <button
              class="btn btn-success"
              type="button"
            >
              ADD
            </button>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="Stocks"
            />
          </td>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="Savings"
            />
          </td>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="25"
            />
          </td>
          <td>
            <button
              class="btn btn-danger"
              type="button"
            >
              DEL
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="Gold"
            />
          </td>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="Savings"
            />
          </td>
          <td>
            <input
              class="form-control"
              placeholder=""
              value="10"
            />
          </td>
          <td>
            <button
              class="btn btn-danger"
              type="button"
            >
              DEL
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
